# 01 Data Exploration

This notebook explores the raw sales dataset for the NorthStar Sales Performance Analytics project.

The goal is to understand the dataset structure, review column names, check data types, identify missing values, and detect duplicate records before performing data cleaning.

In [1]:
import pandas as pd

## 2. Load Dataset

The raw Excel file is stored in the `data/raw` folder.

In [2]:
df = pd.read_excel("../data/raw/Sample - Superstore.xls")

## 3. Preview the Data

Display the first few rows to understand the structure of the dataset.

In [3]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country/Region,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2019-152156,2019-11-08,2019-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420.0,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2019-152156,2019-11-08,2019-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420.0,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2019-138688,2019-06-12,2019-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036.0,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2018-108966,2018-10-11,2018-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311.0,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2018-108966,2018-10-11,2018-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311.0,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


## 4. Dataset Shape

Check the number of rows and columns in the dataset.

In [4]:
df.shape

(9994, 21)

## 5. Column Names

Review all column names in the dataset.

In [5]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country/Region', 'City',
       'State', 'Postal Code', 'Region', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit'],
      dtype='str')

## 6. Dataset Information

Review data types, non-null counts, and memory usage.

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          9994 non-null   int64         
 1   Order ID        9994 non-null   str           
 2   Order Date      9994 non-null   datetime64[us]
 3   Ship Date       9994 non-null   datetime64[us]
 4   Ship Mode       9994 non-null   str           
 5   Customer ID     9994 non-null   str           
 6   Customer Name   9994 non-null   str           
 7   Segment         9994 non-null   str           
 8   Country/Region  9994 non-null   str           
 9   City            9994 non-null   str           
 10  State           9994 non-null   str           
 11  Postal Code     9983 non-null   float64       
 12  Region          9994 non-null   str           
 13  Product ID      9994 non-null   str           
 14  Category        9994 non-null   str           
 15  Sub-Category   

## 7. Summary Statistics

Review summary statistics for numeric columns such as Sales, Quantity, Discount, and Profit.

In [7]:
df.describe()

,Row ID,Order Date,Ship Date,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994,9994,9983.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,2019-04-30 17:41:20.448269,2019-05-04 16:42:15.921553,55245.233297,229.858001,3.789574,0.156203,28.656896
min,1.000000,2017-01-03 00:00:00,2017-01-07 00:00:00,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,2018-05-23 00:00:00,2018-05-27 00:00:00,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,2019-06-26 00:00:00,2019-06-29 00:00:00,57103.000000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,2020-05-14 00:00:00,2020-05-18 00:00:00,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,2020-12-30 00:00:00,2021-01-05 00:00:00,99301.000000,22638.480000,14.000000,0.800000,8399.976000
std,2885.163629,NaN,NaN,32038.715955,623.245101,2.225110,0.206452,234.260108


## 8. Missing Values

Check for missing values in each column.

In [8]:
df.isnull().sum()

Row ID             0
Order ID           0
Order Date         0
Ship Date          0
Ship Mode          0
Customer ID        0
Customer Name      0
Segment            0
Country/Region     0
City               0
State              0
Postal Code       11
Region             0
Product ID         0
Category           0
Sub-Category       0
Product Name       0
Sales              0
Quantity           0
Discount           0
Profit             0
dtype: int64

## 9. Duplicate Rows

Check whether the dataset contains duplicate records.

In [9]:
df.duplicated().sum()

np.int64(0)

## 10. Summary Statistics for All Columns

Display summary statistics for both numeric and non-numeric columns.

In [10]:
df.describe(include="all")

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country/Region,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
count,9994.000000,9994,9994,9994,9994,9994,9994,9994,9994,9994,...,9983.000000,9994,9994,9994,9994,9994,9994.000000,9994.000000,9994.000000,9994.000000
unique,NaN,5009,NaN,NaN,4,793,793,3,1,531,...,NaN,4,1862,3,17,1849,NaN,NaN,NaN,NaN
top,NaN,CA-2020-100111,NaN,NaN,Standard Class,WB-21850,William Brown,Consumer,United States,New York City,...,NaN,West,OFF-PA-10001970,Office Supplies,Binders,Staple envelope,NaN,NaN,NaN,NaN
freq,NaN,14,NaN,NaN,5968,37,37,5191,9994,915,...,NaN,3203,19,6026,1523,48,NaN,NaN,NaN,NaN
mean,4997.500000,NaN,2019-04-30 17:41:20.448269,2019-05-04 16:42:15.921553,NaN,NaN,NaN,NaN,NaN,NaN,...,55245.233297,NaN,NaN,NaN,NaN,NaN,229.858001,3.789574,0.156203,28.656896
min,1.000000,NaN,2017-01-03 00:00:00,2017-01-07 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,...,1040.000000,NaN,NaN,NaN,NaN,NaN,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,NaN,2018-05-23 00:00:00,2018-05-27 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,...,23223.000000,NaN,NaN,NaN,NaN,NaN,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,NaN,2019-06-26 00:00:00,2019-06-29 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,...,57103.000000,NaN,NaN,NaN,NaN,NaN,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,NaN,2020-05-14 00:00:00,2020-05-18 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,...,90008.000000,NaN,NaN,NaN,NaN,NaN,209.940000,5.000000,0.200000,29.364000
max,9994.000000,NaN,2020-12-30 00:00:00,2021-01-05 00:00:00,NaN,NaN,NaN,NaN,NaN,NaN,...,99301.000000,NaN,NaN,NaN,NaN,NaN,22638.480000,14.000000,0.800000,8399.976000


## 11. Unique Values

Determine the number of unique values in each column.

In [11]:
df.nunique()

Row ID            9994
Order ID          5009
Order Date        1236
Ship Date         1334
Ship Mode            4
Customer ID        793
Customer Name      793
Segment              3
Country/Region       1
City               531
State               49
Postal Code        630
Region               4
Product ID        1862
Category             3
Sub-Category        17
Product Name      1849
Sales             6144
Quantity            14
Discount            12
Profit            7545
dtype: int64

## 12. Random Sample

Display a random sample of records to verify that the data looks accurate and consistent throughout the dataset.

A fixed `random_state` is used to ensure the same random sample is generated each time the notebook is run, making the results reproducible and consistent.

```python
df.sample(5, random_state=42)
```

In [12]:
df.sample(5, random_state=42)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country/Region,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
3125,3126,CA-2018-121720,2018-06-11,2018-06-12,First Class,JE-15610,Jim Epp,Corporate,United States,Lakeland,...,33801.0,South,OFF-ST-10003816,Office Supplies,Storage,Fellowes High-Stak Drawer Files,563.808,4,0.2,21.1428
1441,1442,CA-2020-128160,2020-12-19,2020-12-24,Second Class,MM-17920,Michael Moore,Consumer,United States,San Francisco,...,94110.0,West,OFF-BI-10001510,Office Supplies,Binders,Deluxe Heavy-Duty Vinyl Round Ring Binder,36.672,2,0.2,11.4600
4510,4511,CA-2019-119935,2019-11-10,2019-11-14,Standard Class,KM-16225,Kalyca Meade,Corporate,United States,Springfield,...,65807.0,Central,FUR-FU-10001085,Furniture,Furnishings,3M Polarizing Light Filter Sleeves,37.300,2,0.0,17.1580
39,40,CA-2018-117415,2018-12-27,2018-12-31,Standard Class,SN-20710,Steve Nguyen,Home Office,United States,Houston,...,77041.0,Central,FUR-CH-10004218,Furniture,Chairs,"Global Fabric Manager's Chair, Dark Gray",212.058,3,0.3,-15.1470
4509,4510,CA-2020-167003,2020-05-23,2020-05-29,Standard Class,VS-21820,Vivek Sundaresam,Consumer,United States,Los Angeles,...,90036.0,West,FUR-TA-10001520,Furniture,Tables,"Lesro Sheffield Collection Coffee Table, End T...",171.288,3,0.2,-6.4233


## 13. Initial Observations

### Findings

- The dataset contains **9,994** sales records and **21** columns.
- Most columns contain complete data.
- The **Postal Code** column contains **11 missing values**.
- No duplicate records were identified.
- The dataset includes a mix of numeric, categorical, and datetime fields suitable for business analysis.
- Sales and Profit exhibit a wide range of values, indicating potential outliers that will be explored in later phases.

### Conclusion

The dataset is generally well-structured and requires minimal cleaning before exploratory data analysis. The only notable issue identified during exploration is 11 missing values in the Postal Code column, while no duplicate records were detected.

## Purpose of Each Exploration Function

| Function | Question It Answers |
|----------|---------------------|
| `df.head()` | What does the dataset look like? |
| `df.shape` | How big is the dataset (rows and columns)? |
| `df.columns` | What variables (columns) are available? |
| `df.info()` | What data types do I have, and are there any missing values? |
| `df.describe()` | What do the numeric values look like (summary statistics)? |
| `df.isna().sum()` | Are there any missing values in the dataset? |
| `df.duplicated().sum()` | Are there any duplicate rows? |
| `df.nunique()` | How many unique values are in each column? |
| `df.sample()` | Does the data look reasonable throughout the dataset? |